File: js/src/interfaces/IChainApi.ts
import { SignatureCallResult } from './SignatureCallResult';

/**
 * Minimal read-only blockchain API needed by ISignatureVerifier.
 */
export interface IChainApi {
  /**
   * Gets the chain ID.
   */
  getChainId(): number;
  
  /**
   * Returns the contract code at the given address.
   * 
   * @param address The address to check
   * @returns The contract code as a 0x-prefixed hex string
   */
  getCodeAsync(address: string): Promise<string>;
  
  /**
   * Executes the ERC-1271 isValidSignature call via eth_call.
   * Never throws on "execution reverted" - that is surfaced via SignatureCallResult.reverted.
   * 
   * @param address The contract address
   * @param abi The ABI to use
   * @param dataOrHash The data or hash to verify
   * @param signature The signature
   * @returns The result of the call
   */
  callIsValidSignatureAsync(
    address: string,
    abi: any[],
    dataOrHash: Uint8Array,
    signature: Uint8Array
  ): Promise<SignatureCallResult>;
  
  /**
   * Gets the nonce of a Safe.
   * 
   * @param safeAddress The Safe address
   * @returns The nonce
   */
  getSafeNonceAsync(safeAddress: string): Promise<bigint>;
}

File: Circles.Profiles.Interfaces/IChainApi.cs
using System.Numerics;

namespace Circles.Profiles.Interfaces;

/// <summary>
/// Result of an ERC‑1271 <c>isValidSignature</c> pre‑flight.
/// • <see cref="Reverted"/>   – the EVM reverted (i.e. “invalid signature”).  
/// • <see cref="ReturnData"/> – raw ABI‑encoded value (empty when reverted).
/// </summary>
public readonly record struct SignatureCallResult(bool Reverted, byte[] ReturnData);

/// <summary>
/// Minimal read‑only blockchain API needed by <see cref="ISignatureVerifier"/>.
/// </summary>
public interface IChainApi
{
    BigInteger Id { get; }
    
    /// <summary>Returns the contract code at <paramref name="address"/> as a 0x‑prefixed hex string.</summary>
    Task<string> GetCodeAsync(string address, CancellationToken ct = default);

    /// <summary>
    /// Executes the ERC‑1271 <c>isValidSignature</c> call via <c>eth_call</c>.
    /// Never throws on “execution reverted” – that is surfaced via <see cref="SignatureCallResult.Reverted"/>.
    /// </summary>
    Task<SignatureCallResult> CallIsValidSignatureAsync(
        string address,
        string abi,
        byte[] dataOrHash,
        byte[] signature,
        CancellationToken ct = default);
    
    Task<BigInteger> GetSafeNonceAsync(
        string safeAddress, CancellationToken ct = default);
}